substrate/frame/revive/src/exec.rs

```diff
@@ -871,69 +871,100 @@ where
 		read_only: bool,
 		origin_is_caller: bool,
 	) -> Result<Option<(Frame<T>, E)>, ExecError> {
-		let (account_id, contract_info, executable, delegate, entry_point) = match frame_args {
-			FrameArgs::Call { dest, cached_info, delegated_call } => {
-				let contract = if let Some(contract) = cached_info {
-					contract
-				} else {
-					if let Some(contract) =
-						<ContractInfoOf<T>>::get(T::AddressMapper::to_address(&dest))
-					{
+		let (account_id, contract_info, executable, delegate, entry_point, nested_gas) =
+			match frame_args {
+				FrameArgs::Call { dest, cached_info, delegated_call } => {
+					let contract = if let Some(contract) = cached_info {
 						contract
 					} else {
-						return Ok(None);
-					}
-				};
+						if let Some(contract) =
+							<ContractInfoOf<T>>::get(T::AddressMapper::to_address(&dest))
+						{
+							contract
+						} else {
+							return Ok(None);
+						}
+					};
 
-				let (executable, delegate_caller) =
-					if let Some(DelegatedCall { executable, caller, callee }) = delegated_call {
-						(executable, Some(DelegateInfo { caller, callee }))
+					let (executable, delegate_caller, nested_gas) = if let Some(DelegatedCall {
+						executable,
+						caller,
+						callee,
+					}) = delegated_call
+					{
+						(
+							executable,
+							Some(DelegateInfo { caller, callee }),
+							gas_meter.nested(gas_limit),
+						)
+
+					// For the first frame, charge the gas to the frame's nested gas meter
+					// instead of the top-level gas meter, so it's properly associated with
+					// the initial frame when tracing the call.
```

Comment:
Why is this only true for the first frame? Shouldn't we always charge this gas in the nested frame then?

Reply:
the honest answer is:  I tried that first and it broke a few tests. 
Regardless, I think it make sense to charge loading the code from the current frame, and not the child frame.

The initial gas_meter does not belong to any frame, it's just the one we create in bare_call, if we charge from it then we can't associate that gas to any frame, thus the update here

Reply:
> the honest answer is: I tried that first and it broke a few tests 

Can you investigate what broke the tests? I don't think we want a special casing here. It will cause more problems down the line and add some mental load.

Reply:
Okay with the recent changes we consistently  charge the code load token from the nested meter?